cli/src/cast_opts.rs

```diff
@@ -252,19 +253,37 @@ pub struct EthereumOpts {
     pub wallet: Wallet,
 }
 
-// TODO: Improve these so that we return a middleware trait object
 impl EthereumOpts {
-    #[allow(unused)]
-    pub fn provider(&self) -> eyre::Result<Arc<Provider<Http>>> {
-        Ok(Arc::new(Provider::try_from(self.rpc_url.as_str())?))
-    }
+    /// Returns a [`SignerMiddleware`] corresponding to the provided private key, mnemonic or hw
+    /// signer
+    pub async fn signer(&self, chain_id: U256) -> eyre::Result<Option<WalletType>> {
+        let provider = Provider::try_from(self.rpc_url.as_str())?;
```

Comment:
this is a slight overhead since we already have a provider at this point.
we could split this function into

`pub async fn signer(&self, chain_id: U256)` which calls the other one

and 

`pub async fn signer_with(&self, chain_id: U256, provider: Provider<Http>)`

so we can use the provider we already have in the `::SenTx` command and call `signer_with(id, provider.clone())`